File: src/main/java/pw/vodes/rimuru/command/commands/CommandAutomod.java
package pw.vodes.rimuru.command.commands;

import org.apache.commons.lang3.StringUtils;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.event.message.MessageCreateEvent;

import pw.vodes.rimuru.command.Command;
import pw.vodes.rimuru.command.CommandType;
import pw.vodes.rimuru.file.AutoMod;
import pw.vodes.rimuru.file.sub.AutoModFile;
import pw.vodes.rimuru.file.sub.AutoModFile.Punishment;

public class CommandAutomod extends Command {

	public CommandAutomod() {
		super("Automod", new String[] {"automod", "filter"}, CommandType.mod);
		setUsage("`filter \"string to filter\" <kick/ban/timeout> <timeout_minutes>`\n"
				+ "Quotes for Strings with spaces. Quotes can also be escaped.\n"
				+ "Regex fully supported.\n"
				+ "`filter list`\n"
				+ "Lists current filters\n"
				+ "`filter remove <index>`\n"
				+ "Removes filter by index from `list`");
	}

	@Override
	public void run(MessageCreateEvent event) {
		var args = getSplitMessage(event);
		
		if(StringUtils.equalsAnyIgnoreCase(args.get(1), "delete", "remove", "rem", "del") && StringUtils.isNumeric(args.get(2))) {
			try {
				var am = AutoMod.getAutomods().get(Integer.parseInt(args.get(2)));
				AutoMod.remove(am);
				event.getChannel().sendMessage("Removed filter!");
			} catch (Exception e) {
				event.getChannel().sendMessage("No filter with that index found!");
			}
		} else if(StringUtils.equalsAnyIgnoreCase(args.get(1), "list", "l")) {
			if(AutoMod.getAutomods().isEmpty()) {
				event.getChannel().sendMessage("No filters.");
				return;
			}
			event.getChannel().sendMessage(getListEmbed());
			return;
		} else if(!args.get(1).isBlank()) {
			AutoModFile am = new AutoModFile();
			am.filteredSequence = args.get(1);
			am.punishment = Punishment.get(args.get(2));
			am.punishmentVal = StringUtils.isNumeric(args.get(3)) ? args.get(3) : "0";
			AutoMod.add(am);
			event.getChannel().sendMessage("Filter added!");
		} else {
			event.getChannel().sendMessage(getUsage());
			return;
		}
	}
	
	private EmbedBuilder getListEmbed() {
		var embed = new EmbedBuilder().setTitle("Filters");
		for(var am : AutoMod.getAutomods()) {
			var val = am.punishment == null ? "Remove only" : (am.punishment == Punishment.TIMEOUT ? (am.punishment.toString() + " " + am.punishmentVal) : am.punishment.toString());
			embed.addField(am.filteredSequence, val);
		}
		return embed;
	}
}

File: src/main/kotlin/pw/vodes/rimurukt/command/commands/CommandAutomod.kt
package pw.vodes.rimurukt.command.commands

import org.javacord.api.entity.message.MessageFlag
import org.javacord.api.entity.message.embed.EmbedBuilder
import org.javacord.api.interaction.*
import pw.vodes.rimurukt.capitalize
import pw.vodes.rimurukt.command.Command
import pw.vodes.rimurukt.command.CommandType
import pw.vodes.rimurukt.command.Commands
import pw.vodes.rimurukt.reply
import pw.vodes.rimurukt.services.AutoMod
import pw.vodes.rimurukt.services.AutoModEntry
import pw.vodes.rimurukt.services.Punishment
import kotlin.jvm.optionals.getOrNull

class CommandAutomod : Command("Automod", arrayOf("filter", "automod"), CommandType.MOD, "automod") {
    override fun getSlashCommandBuilder(): SlashCommandBuilder? {
        val builder = SlashCommandBuilder()
            .setEnabledInDms(false)
            .setName(slashCommandName)
            .setDescription("Add, remove or list filters.")

        builder.addOption(
            SlashCommandOption.createWithOptions(
                SlashCommandOptionType.SUB_COMMAND, "add", "Add new filter", listOf(
                    SlashCommandOption.createStringOption("Filter", "String or Regex to be filtered", true),
                    SlashCommandOption.createWithChoices(
                        SlashCommandOptionType.STRING, "Punishment", "Type of punishment", true, listOf(
                            SlashCommandOptionChoice.create("None", "None"),
                            SlashCommandOptionChoice.create("Kick", "kick"),
                            SlashCommandOptionChoice.create("Ban", "ban"),
                            SlashCommandOptionChoice.create("Timeout", "timeout")
                        )
                    ),
                    SlashCommandOption.createLongOption("Timeout-Length", "The length of timeout in minutes. Defaults to 1.", false)
                )
            )
        )

        if (AutoMod.automods.isNotEmpty())
            builder.addOption(
                SlashCommandOption.createWithOptions(
                    SlashCommandOptionType.SUB_COMMAND, "remove", "Remove filter", listOf(
                        SlashCommandOption.createWithChoices(
                            SlashCommandOptionType.LONG, "Filter", "Filter to remove", true,
                            AutoMod.automods.mapIndexed { index, entry ->
                                SlashCommandOptionChoice.create(entry.filteredSequence, index.toLong())
                            }
                        )
                    )
                )
            )

        builder.addOption(SlashCommandOption.createSubcommand("list", "List current filters"))

        return builder
    }

    override fun runSlashCommand(interaction: SlashCommandInteraction) {
        val option = interaction.options.getOrNull(0)
        if (option == null) {
            interaction.reply("What, how?", true)
            return
        }
        when (option.name) {
            "add" -> {
                AutoMod.automods.add(
                    AutoModEntry(
                        option.options[0].stringValue.get(),
                        Punishment.get(option.options[1].stringValue.get()),
                        option.options.getOrNull(2)?.longValue?.getOrNull() ?: 1
                    )
                )
                AutoMod.save()
                Commands.updateSlashCommands()
                interaction.reply("Filter added.")
            }

            "remove" -> {
                AutoMod.automods.removeAt(option.options[0].longValue.get().toInt())
                AutoMod.save()
                Commands.updateSlashCommands()
                interaction.reply("Filter removed.")
            }

            "list" -> interaction.createImmediateResponder().addEmbed(listEmbed()).setFlags(MessageFlag.EPHEMERAL).respond()

            else -> interaction.createImmediateResponder().setContent("Somehow passed a non valid action?").setFlags(MessageFlag.EPHEMERAL).respond()
        }
    }

    private fun listEmbed(): EmbedBuilder {
        val embed = EmbedBuilder().setTitle("Filters")
        if (AutoMod.automods.isEmpty())
            embed.setDescription("There are currently no filters.")
        else
            AutoMod.automods.forEach {
                embed.addField(
                    it.filteredSequence,
                    if (it.punishment == Punishment.TIMEOUT)
                        "Timeout (${it.punishmentVal} min)"
                    else
                        it.punishment?.name?.capitalize() ?: "Remove only"
                )
            }
        return embed
    }
}